multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +906,331 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+		for _, address := range subset.Addresses {
+			if address.TargetRef == nil {
+				continue
+			}
+
+			hostName := address.TargetRef.Name
+			hostLocalName := rcsw.mirroredResourceName(hostName)
+			// use a shared informer
+			localService, err := rcsw.localAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(hostLocalName)
+			if err != nil {
+				if !kerrors.IsNotFound(err) {
+					return err
+				}
+
+				// otherwise, create the service
+				localService, err = rcsw.createHeadlessMirroredHost(ctx, address.TargetRef, remoteService, hostLocalName)
+				if err != nil {
+					return err
+				}
+			}
+
+			localServiceNames[hostLocalName] = struct{}{}
+			newAddresses = append(newAddresses, corev1.EndpointAddress{
+				Hostname: hostName,
+				IP:       localService.Spec.ClusterIP,
+			})
+		}
+
+		if len(newAddresses) == 0 {
+			continue
+		}
+		// copy ports, create subset
+		newSubsets = append(newSubsets, corev1.EndpointSubset{
+			Addresses: newAddresses,
+			Ports:     subset.DeepCopy().Ports,
+		})
+	}
+
+	matchLabels := map[string]string{
+		consts.MirroredRootHeadlessLabel: rcsw.mirroredResourceName(remoteService.Name),
+	}
+
+	services, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return err
+	}
+
+	var errors []error
+	for _, service := range services {
+		if _, found := localServiceNames[service.Name]; found {
+			continue
+		}
+		err := rcsw.localAPIClient.Client.CoreV1().Services(service.Namespace).Delete(ctx, service.Name, metav1.DeleteOptions{})
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				errors = append(errors, fmt.Errorf("could not delete Service: %s/%s:%s", service.Namespace, service.Name, err))
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return RetryableError{errors}
+	}
+
+	// Update
+	localEp.Subsets = newSubsets
+	_, err = rcsw.localAPIClient.Client.CoreV1().Endpoints(localEp.Namespace).Update(ctx, localEp, metav1.UpdateOptions{})
+	if err != nil {
+		return RetryableError{[]error{err}}
+	}
+
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) handleHeadlessEndpointsCreated(ctx context.Context, remoteEp *corev1.Endpoints) error {
+	remoteService, err := rcsw.remoteAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(remoteEp.Name)
+	if err != nil {
+		return err
+	}
+	localServiceName := rcsw.mirroredResourceName(remoteService.Name)
+	serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
+
+	// To mirror a headless service, we need to create a clusterIP for each
+	// host in its Endpoints.
+	hosts := make(map[string]struct{})
+	subsetsToCreate := make([]corev1.EndpointSubset, 0, len(remoteEp.Subsets))
+	for _, subset := range remoteEp.Subsets {
+		newAddresses := make([]corev1.EndpointAddress, 0, len(subset.Addresses))
+		for _, addr := range subset.Addresses {
+			// Create annotations for host's service
+			if addr.TargetRef == nil {
```

Comment:
Does there need to be a backing pod?  Can we get all the info we need just from the `addr`?